res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,1949 @@
+                component.group = "[QuickEffectRack1_" + newGroup + "]";
+            }
+
+            component.color = this.groupsToColors[newGroup];
+        });
+    }
+    static groupForNumber(deckNumber) {
+        return "[Channel" + deckNumber + "]";
+    }
+}
+
+class Button extends Component {
+    constructor(options) {
+        super(options);
+        this.off = 0;
+        if (this.longPressTimeOut === undefined) {
+            this.longPressTimeOut = 225; // milliseconds
+        }
+        if (this.inBitLength === undefined) {
+            this.inBitLength = 1;
+        }
+    }
+    output(value) {
+        const brightness = (value > 0) ? this.brightnessOn : this.brightnessOff;
+        this.send(this.color + brightness);
+    }
+}
+
+class PushButton extends Button {
+    constructor(options) {
+        super(options);
+    }
+    input(pressed) {
+        engine.setValue(this.group, this.inKey, pressed);
+    }
+}
+
+class ToggleButton extends Button {
+    constructor(options) {
+        super(options);
+    }
+    input(pressed) {
+        if (pressed) {
+            script.toggleControl(this.group, this.inKey);
+        }
+    }
+}
+
+class PowerWindowButton extends Button {
+    constructor(options) {
+        super(options);
+        this.isLongPressed = false;
+        this.longPressTimer = 0;
+    }
+    input(pressed) {
+        if (pressed) {
+            script.toggleControl(this.group, this.inKey);
+            this.longPressTimer = engine.beginTimer(this.longPressTimeOut, () => {
+                this.isLongPressed = true;
+                this.longPressTimer = 0;
+            }, true);
+        } else {
+            if (this.isLongPressed) {
+                script.toggleControl(this.group, this.inKey);
+            }
+            if (this.longPressTimer !== 0) {
+                engine.stopTimer(this.longPressTimer);
+            }
+            this.longPressTimer = 0;
+            this.isLongPressed = false;
+        }
+    }
+}
+
+class PlayButton extends ToggleButton {
+    constructor(options) {
+        super(options);
+        this.inKey = "play";
+        this.outKey = "play_indicator";
+        this.outConnect();
+    }
+}
+
+class CueButton extends PushButton {
+    constructor(options) {
+        super(options);
+        this.outKey = "cue_indicator";
+        this.outConnect();
+    }
+    unshift() {
+        this.inKey = "cue_default";
+    }
+    shift() {
+        this.inKey = "start_stop";
+    }
+}
+
+class Encoder extends Component {
+    constructor(options) {
+        super(options);
+        this.lastValue = null;
+    }
+    isRightTurn(value) {
+        // detect wrap around
+        const oldValue = this.lastValue;
+        this.lastValue = value;
+        if (oldValue === this.max && value === 0) {
+            return true;
+        }
+        if (oldValue === 0 && value === this.max) {
+            return false;
+        }
+        return value > oldValue;
+    }
```

Comment:
I am not sure if the S4 Mk3 encoders are relative or absolute. I think relative. But either way this function is helpful to handle wrap arounds.

Reply:
but it only handles wrap arounds if the value increased monotonically with a stepsize of 0. So we should try to tidy it up and generalize it more. But thats difficult for me right now because I don't know what input it expects.

Reply:
I am not aware of HID controllers with encoders that work any other way. YAGNI. If some controller really does need it, the script can implement its own Encoder class.

Reply:
Well I'm aware of controllers that generally do this. For example the jogwheels on mine are essentially encoders that sum up multiple midi messages into one  message with the sum of the values. I believe my browse-encoder does something similar.

Reply:
I doubt it would be a good idea to conflate encoders and jog wheels.

Reply:
That is debatable.

Reply:
Moreover, looking at above code, an encoder that reports absolute values is exactly the usecase where a signed integer is appropriate.